src/modal/index.tsx

```diff
@@ -42,37 +42,58 @@ class Modal extends Component <ModalProps> {
     displayDimmer: true,
   }
 
+  constructor(props: ModalProps) {
+    super(props)
+    this.portalNode = document.createElement('div')
+    document.body.appendChild(this.portalNode)
+  }
+
   componentDidMount() {
-    if (this.props.closeOnEsc && canUseEventListeners) {
-      document.addEventListener('keydown', this.handleKeydown)
+    if (this.props.isOpen) {
+      this.addListeners()
+    }
+  }
+
+  componentDidUpdate(prevProps: ModalProps) {
+    if (!prevProps.isOpen && this.props.isOpen) {
+      this.addListeners()
     }
-    if (this.props.closeOnOutsideClick && canUseEventListeners) {
-      document.addEventListener('mouseup', this.handleOutsideMouseClick)
-      document.addEventListener('touchstart', this.handleOutsideMouseClick)
+
+    if (!this.props.isOpen && prevProps.isOpen) {
+      this.removeListeners()
     }
   }
 
   componentWillUnmount() {
-    if (this.props.closeOnEsc && canUseEventListeners) {
-      document.removeEventListener('keydown', this.handleKeydown)
+    this.removeListeners()
+    document.body.removeChild(this.portalNode)
+    this.portalNode = null
+  }
+
+  addListeners() {
+    if (this.props.isOpen && canUseEventListeners) {
+      if (this.props.closeOnEsc) {
+        document.addEventListener('keydown', this.handleKeydown)
+      }
+
+      if (this.props.closeOnOutsideClick) {
+        document.addEventListener('mouseup', this.handleOutsideMouseClick)
+        document.addEventListener('touchstart', this.handleOutsideMouseClick)
+      }
     }
-    if (this.props.closeOnOutsideClick && canUseEventListeners) {
+  }
+
+  removeListeners() {
+    if (!this.props.isOpen && canUseEventListeners) {
+      document.removeEventListener('keydown', this.handleKeydown)
```

Comment:
Should we have an object referencing all the events? So that if we add a listener we don't forget to remove it